contrib/win32/win32compat/w32-sshfileperm.c

```diff
@@ -247,15 +253,96 @@ check_secure_folder_permission(const wchar_t* path_utf16, int read_ok)
+				resolved_trustee_domain, &resolved_trustee_domain_len, &resolved_trustee_type) != 0) {
+				if (is_first) {
+					_snwprintf_s(log_msg, log_msg_len, _TRUNCATE, L"%ls\\%ls", resolved_trustee_domain, resolved_trustee);
+					is_first = FALSE;
+				}
+				else {
+					size_t currentLength = wcslen(log_msg);
+					size_t userLength = resolved_trustee_domain_len + 1 + resolved_trustee_len + 2; // +1 for '\\' and +2 for ', '
+					if (wcslen(log_msg) + userLength + 1 > log_msg_len) { // +1 for null terminator
+						log_msg_len *= 2;
+						wchar_t* temp_log_msg = (wchar_t*)malloc(log_msg_len * sizeof(wchar_t));
+						if (temp_log_msg == NULL) {
+							break;
+						}
+						wcscpy_s(temp_log_msg, log_msg_len, log_msg);
+						if (log_msg)
+							free(log_msg);
+						log_msg = temp_log_msg;
+					}
+					_snwprintf_s(log_msg + currentLength, log_msg_len - currentLength, _TRUNCATE, 
+						L", %ls\\%ls", resolved_trustee_domain, resolved_trustee);
+				}
+			}
 		}
 	}
+
+	if (ret != 0) {
+		log_folder_permissions_message(path_utf16, log_msg);
+	}
 cleanup:
 	if (bad_user)
 		LocalFree(bad_user);
+	if (log_msg)
+		free(log_msg);
 	if (pSD)
 		LocalFree(pSD);
 	if (ti_sid)
 		free(ti_sid);
-	return ret;
+}
+
+/* Helper function used by check_secure_folder_permission */
+void log_folder_permissions_message(const wchar_t* path_utf16, wchar_t* log_msg) {
+	log_on_stderr = 0;
+
+	PSID adminSid = NULL;
+	WCHAR adminName[UNLEN + 1];
+	WCHAR adminDomain[DNLEN + 1];
+	DWORD adminNameSize = UNLEN + 1;
+	DWORD adminDomainSize = DNLEN + 1;
+	DWORD adminSidSize = SECURITY_MAX_SID_SIZE;
+	PSID systemSid = NULL;
+	WCHAR systemName[UNLEN + 1];
+	WCHAR systemDomain[DNLEN + 1];
+	DWORD systemNameSize = UNLEN + 1;
+	DWORD systemDomainSize = DNLEN + 1;
+	DWORD systemSidSize = SECURITY_MAX_SID_SIZE;
+	SID_NAME_USE sidType;
+
+	adminSid = (PSID)malloc(SECURITY_MAX_SID_SIZE);
+	if (log_msg != NULL && adminSid != NULL &&
+		CreateWellKnownSid(WinBuiltinAdministratorsSid, NULL, adminSid, &adminSidSize) != 0 &&
+		LookupAccountSidW(NULL, adminSid, adminName, &adminNameSize, adminDomain, &adminDomainSize, &sidType) != 0) {
+		systemSid = (PSID)malloc(SECURITY_MAX_SID_SIZE);
+		if (systemSid != NULL &&
+			CreateWellKnownSid(WinLocalSystemSid, NULL, systemSid, &systemSidSize) != 0 &&
+			LookupAccountSidW(NULL, systemSid, systemName, &systemNameSize, systemDomain, &systemDomainSize, &sidType) != 0) {
+			logit("For '%S' folder, write access is granted to the following users: %S. "
+				"Consider reviewing users to ensure that only %S\\%S, and the %S\\%S group, and its members, have write access.", 
+				path_utf16, log_msg, systemDomain, systemName, adminDomain, adminName);
+			log_on_stderr = 1;
+		}
+	}
+
+	if (log_on_stderr == 0) {
+		/* log generic warning message in unlikely case that lookup for either well-known SID fails or user list is empty */
+		logit("for '%S' folder, consider downgrading permissions for any users with unnecessary write access.", path_utf16);
```

Comment:
Is `logit()` an upstream function or only in the Windows code?  If the latter, seems better to have a wrapper function that only logs to ETW as that seems like the intent?

Reply:
It's the former - declared/defined in log.h/log.c.